profiler/profiler.go

```diff
@@ -78,12 +78,25 @@ func newProfiler(opts ...Option) (*profiler, error) {
 	if os.Getenv("DD_PROFILING_WAIT_PROFILE") != "" {
 		cfg.addProfileType(expGoroutineWaitProfile)
 	}
-	if cfg.apiKey != "" {
+	// Agentless upload is disabled by default as of v1.30.0, but
+	// WithAgentlessUpload can be used to enable it for testing and debugging.
+	if cfg.agentless {
 		if !isAPIKeyValid(cfg.apiKey) {
-			return nil, errors.New("API key has incorrect format")
+			return nil, errors.New("profiler.WithAgentlessUpload requires a valid API key. Use profiler.WithAPIKey or the DD_API_KEY env variable to set it")
 		}
+		// Always warn people against using this mode for now. All customers should
+		// use agent based uploading at this point.
+		log.Warn("profiler.WithAgentlessUpload is currently for internal usage only and not officially supported.")
 		cfg.targetURL = cfg.apiURL
 	} else {
+		// Historically people could use an API Key to enable agentless uploading.
+		// As of v1.30.0 customers the default behavior is to use agent based
+		// uploading regardless of the presence of an API key. So if we see an API
+		// key configured, we warn the customers that this is probably a
+		// misconfiguration.
+		if cfg.apiKey != "" {
+			log.Warn("You are currently setting profiler.WithAPIKey or the DD_API_KEY env variable, but as of dd-trace-go v1.30.0 this value is getting ignored by the profiler. Please see the profiler.WithAPIKey go docs and verify that your integration is still working. If you can't remove DD_API_KEY from your environment, you can use WithAPIKey(\"\") to silence this warning.")
```

Comment:
Nit: This might be too "aggressive", specially on those cases where people have no control. You do mention they can remove the warning by changing but is it gaining us anything?

Reply:
Also, why are we warning when agentless is still a legitimate use case?

Reply:
> Nit: This might be too "aggressive" as it's rather big, specially on those cases where people have no control. You do mention they can remove the warning by changing but is it gaining us anything? Anyway not a big deal for me, I just know that some tracers care about reducing verbosity.

This PR introduces a backwards incompatible change in a minor release of the profiling library. Arguably a big warning is warranted. But if you think dropping the bit about removing the warning would help, I'd be okay with that. Just let me know.

I'm also okay to change the entire message if you have a different proposal, just let me know.

> Also, why are we warning when agentless is still a legitimate use case?

There is no legitimate production use case at this point. The main use case is for our own internal testing, and perhaps to help customers troubleshoot connectivity issues. But none of that code should ever run in production, so I think the warning makes sense?

Reply:
> There is no legitimate production use case at this point.
Going to push back on that a bit.  I can think of a number of use cases that don't run afoul of billing issues where a customer might not want to run the agent.  Maybe I'm just not informed.

Reply:
Legitimate production use cases aside (I'm going to kind of disagree with that assertion), I do not think this should be a warning.  Documented or not, the client does support running agentless, even if it is no longer the default.

Reply:
@pmbauer perhaps I don't fully understand the use cases you have in mind. Let's continue the discussion in the [google doc thread](https://docs.google.com/document/d/172cmlRA_GTSu2OjC_FL5PzO6EjCpRH5ddU1nP08-yF0/edit?disco=AAAAIS2SYUg) and make sure we're aligned before continuing the code related discussions here.

Reply:
Replied in the doc.  The question of billing for profiles not submitted via the agent is a product and business decision.  That in no way makes the production uses for agentless uploads illegitimate from a customer perspective.

Reply:
While I do feel strongly about the business decision to force profiles through the agent (it's bad for customers and bad for business long term) I don't feel strongly about the warning.  Feel free to ship.